parsely/src/main/java/com/parsely/parselyandroid/InMemoryBuffer.kt

```
@@ -0,0 +1,43 @@
+package com.parsely.parselyandroid
+
+import kotlin.time.Duration.Companion.seconds
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+internal class InMemoryBuffer(
+    private val coroutineScope: CoroutineScope,
+    private val localStorageRepository: LocalStorageRepository,
+    private val onEventAddedListener: () -> Unit,
+) {
+
+    private val mutex = Mutex()
```

To test if this mutual exclusion works correctly, comment this, usages in `init` and `add` and then run `stressTest`. It should fail with unexpected result - sometimes recorded events are smaller, sometimes bigger.

---

Thanks for the instructions! 🥇 

FYI: I run that `stressTest` 3 times after removing `mutex` and everything was still working as expected on my side (testing on actual device). I then also triggered the whole `./gradlew connectedDebugAndroidTest` from CLI and it still worked as expected, everything was successful. 🤷

---

UPDATE: I then also removed the `mutex` on the repository level and run the `stressTest` 3 times, along with running the whole `./gradlew connectedDebugAndroidTest` from CLI. Running the `stressTest` succeeded every time, but running the whole `./gradlew connectedDebugAndroidTest` from CLI indeed failed (both times I tried):

```
com.parsely.parselyandroid.FunctionalTests > stressTest[Pixel 4 - 13] FAILED 
        java.lang.AssertionError:
        Expected size: 500 but was: 1000 in:
Tests on Pixel 4 - 13 failed: There was 1 failure(s).
```

PS: 
1. I then uncommented the `mutex` on the repository level and run whole `./gradlew connectedDebugAndroidTest` from CLI again and it started failing as well. 👍 
2. I then, uncommented the `mutex` on the buffer level and  run whole `./gradlew connectedDebugAndroidTest` from CLI again and it now succeeded. ✅ 

I guess there exist some flakiness but it seems that everything is working as expected after all. Just wanted to document my testing here, just in case it somehow helps you.

---

Thanks for sharing! I think for some reason, they're more likely to fail without `mutex` on emulator. In any case, one can also remove the `delay` to the `InMemoryBuffer#init` - this would make the test fail more certain.

---

> I think for some reason, they're more likely to fail without mutex on emulator. 

Yea, maybe! 🤷 

> In any case, one can also remove the delay to the InMemoryBuffer#init - this would make the test fail more certain.

True, good tip! 💯